src/repr/src/timestamp.rs

```diff
@@ -182,6 +198,54 @@ impl Timestamp {
     pub fn step_back(&self) -> Option<Self> {
         self.checked_sub(1)
     }
+
+    /// Rounds up the timestamp to the time of the next refresh, according to the given periodic
+    /// refresh schedule.
+    ///
+    /// Returns None if an overflow happens.
+    pub fn round_up(
+        &self,
+        RefreshEvery {
+            interval,
+            aligned_to,
+        }: &RefreshEvery,
+    ) -> Option<Self> {
+        // Planning ensured that
+        // - interval.months == 0, so we don't need to deal with months being of variable size.
+        // - The interval can be max 27 days, so the cast to u64 won't overflow.
+        // - The interval is positive, so the cast to u64 won't underflow.
+        assert_eq!(interval.months, 0);
+        let interval: u64 = interval.as_milliseconds().try_into().unwrap();
+        // Rounds up `x` to the nearest multiple of `interval`.
+        let round_up_to_multiple_of_interval = |x: u64| -> Option<u64> {
+            assert_ne!(x, 0);
+            (((x - 1) / interval).checked_add(1)?).checked_mul(interval)
+        };
+        // Rounds down `x` to the nearest multiple of `interval`.
+        let round_down_to_multiple_of_interval = |x: u64| -> u64 { x / interval * interval };
+        let result = if self > aligned_to {
+            Self {
+                internal: aligned_to
+                    .internal
+                    .checked_add(round_up_to_multiple_of_interval(
+                        self.internal - aligned_to.internal,
+                    )?)?,
+            }
+        } else {
+            // Note: `self == aligned_to` has to be handled here, because in the other branch
+            // `x - 1` would underflow.
+            //
+            // Also, no need to check for overflows here, since all the numbers are either between
+            // `self` and `aligned_to`, or not greater than `aligned_to.internal - self.internal`.
+            Self {
+                internal: aligned_to.internal
+                    - round_down_to_multiple_of_interval(aligned_to.internal - self.internal),
+            }
+        };
+        assert!(result.internal >= self.internal);
+        assert!(result.internal - self.internal <= interval);
+        Some(result)
```

Comment:
If we are worried about the performance of this in the hot persist_sink path, how about making these `assert`s (non-logging) soft-asserts?

Reply:
For now, I'd sleep better if these are always-on asserts. They should be negligible compared to all the other things  going on in this function (or, not to mention, in the entire system). E.g. 64-bit integer division is around 15 clock cycles on modern Intel CPUs.

(An even bigger thing is the `interval.as_milliseconds()` call at the beginning of the function, which involves several integer divisions. I think I'll move the `as_milliseconds()` call into the creation of `RefreshEvery` from this function.)

Reply:
Fine by me! In this case it would be good to have a `TODO` comment to downgrade the asserts once we have built some confidence in the code.

Reply:
Ok, done